crates/sui-storage/src/lock_service.rs

```diff
@@ -241,24 +251,34 @@ impl LockServiceImpl {
 
         for (i, lock) in locks.iter().enumerate() {
             // The object / version must exist, and therefore lock initialized.
-            let lock = lock.ok_or(SuiError::TransactionLockDoesNotExist)?;
-
-            if let Some(previous_tx_digest) = lock {
-                // Lock already set to different transaction
-                if previous_tx_digest != tx_digest {
+            let lock = lock.as_ref().ok_or(SuiError::TransactionLockDoesNotExist)?;
+
+            if let Some(LockInfo {
+                epoch: previous_epoch,
+                tx_digest: previous_tx_digest,
+            }) = lock
+            {
+                // Lock already set to different transaction from the same epoch.
+                // If the lock is set in a previous epoch, it's ok to override it.
+                if previous_epoch == &epoch && previous_tx_digest != &tx_digest {
```

Comment:
Shouldn't we simply clear all locks upon epoch change?
What if the lock was set in an epoch older than the previous one?

Reply:
> What if the lock was set in an epoch older than the previous one?

It shouldn't matter, as long as the epoch is different, we can just override it. The effect of overriding should be the same as clearing locks at epoch change.

> Shouldn't we simply clear all locks upon epoch change?

If it is easy to do so we should. However the current lock table serves two purposes, one is an indication that locks are initialized, and one is that it may be locked by a transaction. To clear all entries that are locked by a transaction, we would have to iterate through the entire table. One way to do this may be we split it into two tables. But I don't see a need for it if overriding just works.